packages/block-library/src/block/edit.js

```diff
@@ -27,14 +29,86 @@ import {
 	useBlockProps,
 	Warning,
 	privateApis as blockEditorPrivateApis,
+	store as blockEditorStore,
 } from '@wordpress/block-editor';
-import { useRef, useMemo } from '@wordpress/element';
+import { getBlockSupport } from '@wordpress/blocks';
 
 /**
  * Internal dependencies
  */
 import { unlock } from '../lock-unlock';
 
+function hasAttributeSynced( block ) {
+	return (
+		!! getBlockSupport( block.name, '__experimentalConnections', false ) &&
+		!! block.attributes.connections?.attributes &&
+		Object.values( block.attributes.connections.attributes ).some(
+			( connection ) => connection.source === 'pattern_attributes'
+		)
+	);
+}
+function getAttributeSynced( block ) {
+	const attributes = {};
+	for ( const [ attribute, connection ] of Object.entries(
+		block.attributes.connections.attributes
+	) ) {
+		if ( connection.source !== 'pattern_attributes' ) continue;
+		attributes[ attribute ] = connection.value;
+	}
+	return attributes;
+}
+
+const updateBlockAttributes =
+	( patternClientId ) =>
+	( clientIds, attributes, uniqueByBlock = false ) =>
+	( { select, dispatch } ) => {
+		const updates = {};
+		for ( const clientId of [].concat( clientIds ) ) {
+			const attrs = uniqueByBlock ? attributes[ clientId ] : attributes;
+			const parentPattern = select.getBlock( patternClientId );
```

Comment:
Have you considered reading the `clientId` of the pattern by searching for the first parent of the `core/block` type? This way, it's going to be easier to move the logic outside of the subregistry. It can be done later if that is the best way in the current setup.